### PR TITLE
ingress V1: implement status sync

### PR DESCRIPTION
**Pull request by mflendrich**
This PR makes `statusSync` update the `status` of Ingress v1 resources in Store just like it does for Ingress v1beta1.
Includes a test (equivalent to the v1beta1 test).

Notes to reviewers:
- This implementation (and test) is an almost direct copy of the preexisting v1beta1 code, including todos and bugs (most notably #829). For review comments regarding v1 functionality, please think whether the same problem applies to v1beta1. For such (preexisting) problems, use your best judgment to tell if a fix should belong in scope of this PR (which I define as "introduce v1 syncing") or not (and be therefore reported as an issue affecting both v1beta1 and v1 to be fixed later)
- Consider not squashing when merging

### CODE REVIEW

File: internal/ingress/status/status.go [hunk 1]
```diff
@@ -363,6 +367,14 @@ func (s *statusSync) runUpdate(ctx context.Context, ing *networkingv1beta1.Ingre
 		}
 
 		switch s.IngressAPI {
+		case utils.NetworkingV1:
+			// I expect this case to never happen, because if s.IngressAPI == NetworkingV1, then I expect Store to have only
+			// v1 ingresses (and no v1beta1 ingresses). If Store happens to have a v1beta1 Ingress nonetheless, I'm choosing
+			// not to drop it, but to log a warning and talk networking.k8s.io/v1beta1 (as opposed to extensions/v1beta1)
+			// because a v1-supporting Kubernetes API is more likely to support the former than the latter.
+			logger.Warnf("statusSync got an unexpected v1beta1 Ingress when it expected v1")
+			fallthrough
```
Comment (rainest) [inline internal/ingress/status/status.go:376]:
```suggestion
		case utils.NetworkingV1:
			// I expect this case to never happen, because if s.IngressAPI == NetworkingV1, then I expect Store to have only
			// v1 ingresses (and no v1beta1 ingresses). If Store happens to have a v1beta1 Ingress nonetheless, I'm choosing
			// not to drop it, but to log a warning and talk networking.k8s.io/v1beta1 (as opposed to extensions/v1beta1)
			// because a v1-supporting Kubernetes API is more likely to support the former than the latter.
			logger.Debugf("statusSync got an unexpected v1beta1 Ingress when it expected v1")
			fallthrough
```

Is there any K8S-level restriction or expectation that this is the case? I don't think there is, and I wouldn't be surprised if we do encounter environments with a mix of v1 and v1beta1 in practice: that wouldn't be unusual if, say, there are multiple teams who share the controller for their applications, but don't otherwise interact much, and migrate from old to new at their own pace.

That scenario is actually quite common; we have many users where an infrastructure team manages the controller and K8S cluster on behalf of application developer teams, and the application developer teams are largely independent of one another.

We can arguably even hold off on adding a warning at this point--I'd argue in favor of that unless we expect some other issue to arise from using both at once, since the error could be confusing/imply that users _need_ to only use one or the other.

1.19 does deprecate v1beta1, and K8S plan to remove support in 1.22, but I expect we'll continue support for older versions for a while. Depending on our timelines, I'd say drop this to debug for now, and raise to warn when we're planning to drop support for versions <1.22.

**Review by rainest — DISMISSED** [review]

File: internal/ingress/status/status.go (hunk 1 above)
Comment (mflendrich) [reply internal/ingress/status/status.go:376]: This warning has nothing to do with resource versions being used by the end user.

Long story short: KIC negotiates Ingress API version with the k8s apiserver. Since then, all resources KIC receives (and updates in k8s), are in the negotiated apiversion. This is completely independent from what the user uploads to kubernetes.

The way it works is the following: Whenever anyone uploads an Ingress to Kubernetes, apiserver converts it (possibly no-op) to the apiversion which it is configured to store Ingresses in (determined by the version of kubernetes). Then, when talking with clients (kubectl being one, KIC being another), Kubernetes apiserver converts again (possibly no-op) that stored apiversion to the apiversion requested by the client.

After #826, KIC negotiates the apiVersion it uses to talk to the apiserver depending on apiserver's capabilities. Since then KIC expects **all resources** it sees to be in that specific apiversion (even if they were uploaded to kubernetes by the user in another version).

This log print (one referenced by this comment) happens when the expectation above is broken (by some bug in KIC), not when 
>  we do encounter environments with a mix of v1 and v1beta1

In other words, this log print does not indicate that the user uploaded some v1 ingresses and some v1beta1 ingresses, but that there's a bug in KIC. Which means that it says "uh-oh, KIC authors think that this condition is a serious problem, because KIC should **never** encounter this situation" and therefore should stay at warning level.

[This workshop from KubeCon NA 2019](https://static.sched.com/hosted_files/kccncna19/e8/KubeCon%20NA%2019_%20Multi-version%20CRDs.pdf) is very thorough on the matter of modern apiversion management as of k8s 1.16 and beyond.

File: go.mod [hunk 2]
```diff
@@ -13,6 +13,7 @@ require (
 	github.com/kong/go-kong v0.13.0
 	github.com/lithammer/dedent v1.1.0
 	github.com/mitchellh/mapstructure v1.3.1
+	github.com/pkg/errors v0.9.1
```
Comment (hbagdi) [inline go.mod:16]: Please remove this dependency.

**Review by hbagdi — APPROVED** [review]
Please merge the feat/ingressv1-informers-store PR and then rebase and merge this one.
Consider this approved even after the rebase. Thanks for working on this.

File: internal/ingress/status/status.go (hunk 1 above)
Comment (rainest) [reply internal/ingress/status/status.go:376]: Cool, that makes sense, same as the coercion performed for some other resource (forget what, some sort of validation thing) I'd seen for resources created by the Helm chart and chalked off as a "huh, that's weird, but okay" implementation thing--thanks for the talk notes, will have to read that as I was stuck staffing our booth for most of that KubeCon.

File: go.mod (hunk 2 above)
Comment (mflendrich) [reply go.mod]: Done.

**Review by hbagdi — APPROVED** [review]
